libraries/standard/mqtt/src/mqtt.c

```diff
@@ -1,19 +1,103 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <string.h>
+
 #include "mqtt.h"
 
 void MQTT_Init( MQTTContext_t * const pContext,
                 const MQTTTransportInterface_t * const pTransportInterface,
                 const MQTTApplicationCallbacks_t * const pCallbacks,
-                const MQTTFixedBuffer_t * const pTxBuffer,
-                const MQTTFixedBuffer_t * const pRxBuffer )
+                const MQTTFixedBuffer_t * const pNetworkBuffer )
 {
+    memset( pContext, 0x00, sizeof( MQTTContext_t ) );
 
+    pContext->connectStatus = MQTTNotConnected;
+    pContext->pTransportInterface = pTransportInterface;
+    pContext->pCallbacks = pCallbacks;
+    pContext->pNetworkBuffer = pNetworkBuffer;
+
+    /* Zero is not a valid packet ID per MQTT spec. Start from 1. */
+    pContext->nextPacketId = 1;
 }
 
 MQTTStatus_t MQTT_Connect( MQTTContext_t * const pContext,
                            const MQTTConnectInfo_t * const pConnectInfo,
+                           const MQTTPublishInfo_t * const pWillInfo,
                            bool * const pSessionPresent )
 {
-    return MQTTSuccess;
+    size_t remainingLength, packetSize;
+    int32_t bytesSent;
+    uint32_t sendTime;
+    MQTTPacketInfo_t connack;
+
+    MQTTStatus_t status = MQTT_GetConnectPacketSize( pConnectInfo,
+                                                     pWillInfo,
+                                                     &remainingLength,
+                                                     &packetSize );
+
+    if( status == MQTTSuccess )
+    {
+        status = MQTT_SerializeConnect( pConnectInfo,
+                                        pWillInfo,
+                                        remainingLength,
+                                        pContext->pNetworkBuffer );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        sendTime = pContext->pCallbacks->getTime();
+
+        bytesSent= pContext->pTransportInterface->send( pContext->pTransportInterface->networkContext,
+                                                        pContext->pNetworkBuffer->pBuffer,
+                                                        packetSize );
+
+        if( ( bytesSent > 0 ) && ( ( size_t ) bytesSent == packetSize ) )
+        {
+            pContext->lastPacketTime = sendTime;
+        }
+        else
+        {
+            status = MQTTSendFailed;
+        }
+    }
+
+    if( status == MQTTSuccess )
+    {
+        status = MQTT_GetIncomingPacket( pContext->pTransportInterface->recv,
+                                         pContext->pTransportInterface->networkContext,
+                                         &connack );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        status = MQTT_DeserializeAck( &connack, NULL, pSessionPresent );
```

Comment:
Does this call check that CONNACK is received and not any other ACK? The flow is a bit unclear here - I think ```connack``` should be renamed to ```incomingPacket```.  And then somewhere we need to check that a valid CONNACK is received.

Reply:
For send/recv: Lightweight needs recv (to read the packet off the network) but doesn't need send.  Only the full API, which transmits packets, needs send.

Changed names and added check for CONNACK.